src/arena.c

```diff
@@ -1939,6 +1943,58 @@ arena_new(tsdn_t *tsdn, unsigned ind, extent_hooks_t *extent_hooks) {
 	return NULL;
 }
 
+arena_t *
+arena_choose_huge(tsd_t *tsd) {
+	/* huge_arena_ind can be 0 during init (will use a0). */
+	if (huge_arena_ind == 0) {
+		assert(!malloc_initialized());
+	}
+
+	arena_t *huge_arena = arena_get(tsd_tsdn(tsd), huge_arena_ind, false);
+	if (huge_arena == NULL) {
+		/* Create the huge arena on demand. */
+		assert(huge_arena_ind != 0);
+		huge_arena = arena_get(tsd_tsdn(tsd), huge_arena_ind, true);
+		if (huge_arena == NULL) {
+			return NULL;
+		}
+		/*
+		 * Purge eagerly for huge allocations, because: 1) number of
+		 * huge allocations is usually small, which means ticker based
+		 * decay is not reliable; and 2) less immediate reuse is
+		 * expected for huge allocations.
+		 */
+		if (arena_dirty_decay_ms_default_get() > 0) {
+			arena_dirty_decay_ms_set(tsd_tsdn(tsd), huge_arena, 0);
+		}
+		if (arena_muzzy_decay_ms_default_get() > 0) {
+			arena_muzzy_decay_ms_set(tsd_tsdn(tsd), huge_arena, 0);
+		}
+	}
+
+	return huge_arena;
+}
+
+bool
+arena_init_huge(void) {
+	bool huge_enabled;
+
+	/* The threshold should be large size class. */
+	if (opt_huge_threshold > LARGE_MAXCLASS ||
+	    opt_huge_threshold < LARGE_MINCLASS) {
```

Comment:
Similarly, if the threshold is smaller than LARGE_MINCLASS, it seems that we ought to bump it up to LARGE_MINCLASS, not make it unreasonably high.

Reply:
Oh I basically treated these two cases as invalid input, and disabled the feature there. The extreme high value is to make sure the use-huge-arena-or-not check always fail.

Do you think we should clip the value to the acceptable range?

Reply:
I think on the high end it doesn't matter (if you say "I want a huge arena, but only for handling requests bigger than 279 exabytes", then you deserve whatever dumb outcomes you get). But on the low end, I think having a cutoff that's resilient to changes across our decisions of what counts as large make sense.

Or if the user puts in invalid limits and we decide they're confused, couldn't we disable the feature via the setting (instead of by giving it nonsensical limits)?

Reply:
Oh when they put invalid input, there is warning from the config checking complaining about invalid values. The nonsensical limits (huge_threshold) is actually an internal value; the "official" reported value is opt_huge_threshold which is set to 0 (when we document the feature I'll make sure it emphasizes on 0 as disabled.

A slight reason I wanted this internal nonsensical value, is to have `if (unlikely(size >= huge_threshold))` as the check on `arena_choose`, instead of having to add an additional `&& huge_enabled)`. I feel it doesn't do much damage as the value is completely internal and not reported anywhere.

Reply:
OK, sounds reasonable.

Reply:
If it helps a bit, let me rename huge_threshold to `huge_threshold_internal`.

Reply:
Meh, I don't think that's necessary.